tests/test_multi_maintenance.py

```diff
@@ -0,0 +1,199 @@
+    node2.stop_postgres()
+
+    # assigned and goal state must be the same
+    assert node1.wait_until_state(target_state="join_primary")
+
+    print("Calling pgautofailover.failover() on the monitor")
+    monitor.failover()
+    assert node3.wait_until_state(target_state="primary")
+    assert node1.wait_until_state(target_state="secondary")
+
+    print("Disabling maintenance on node2, should connect to the new primary")
+    node2.disable_maintenance()
+
+    # allow manual checking of primary_conninfo
+    print("current primary is node3 at %s" % str(node3.vnode.address))
+
+    if node2.pgmajor() < 12:
+        fn = "recovery.conf"
+    else:
+        fn = "postgresql-auto-failover-standby.conf"
+
+    fn = os.path.join(node2.datadir, fn)
+    print("%s:\n%s" % (fn, open(fn).read()))
+
+    assert node2.wait_until_state(target_state="secondary")
+
+    assert node1.has_needed_replication_slots()
+    assert node2.has_needed_replication_slots()
+    assert node3.has_needed_replication_slots()
+
+def test_007_read_from_new_primary():
+    results = node3.run_sql_query("SELECT * FROM t1")
+    assert results == [(1,), (2,), (3,), (4,)]
+
+def test_008a_node1_to_maintenance():
+    print()
+    assert node3.wait_until_state(target_state="primary")
+
+    print("Enabling maintenance on node1")
+    node1.enable_maintenance()
+
+@raises(Exception)
+def test_008b_node2_to_maintenance_fails():
+    # node3 is the current primary
+    assert node3.get_number_sync_standbys() == 1
+
+    # should fail, because then we would not have enough standby nodes
+    print("Enabling maintenance on node2")
+    node2.enable_maintenance()
+
+def test_009_disable_maintenance():
+    print("Disabling maintenance on node1")
+    node1.disable_maintenance()
+
+    assert node3.wait_until_state(target_state="primary")
+
+def test_010_set_number_sync_standby_to_zero():
+    assert node3.set_number_sync_standbys(0)
+    eq_(node3.get_number_sync_standbys(), 0)
+
+def test_011_all_to_maintenance():
+    print()
+    assert node3.wait_until_state(target_state="primary")
+
+    print("Enabling maintenance on node1")
+    node1.enable_maintenance()
+
+    # now we can, because we don't care about having any standbys
+    print("Enabling maintenance on node2")
+    node2.enable_maintenance()
+
+    assert node3.wait_until_state(target_state="wait_primary")
+
+    # also let's see synchronous_standby_names here
+    print("Monitor: %s" % node3.get_synchronous_standby_names())
+    print("Node 3:  %s" %
+          node3.run_sql_query("show synchronous_standby_names")[0][0])
+
+def test_012_can_write_during_maintenance():
+    node3.run_sql_query("INSERT INTO t1 VALUES (5), (6)")
+    node3.run_sql_query("CHECKPOINT")
+
+def test_014_disable_maintenance():
+    print("Disabling maintenance on node2")
+    node2.disable_maintenance()
+
+    print("Disabling maintenance on node1")
+    node1.disable_maintenance()
+
+    print("Set number_sync_standbys to 1 again")
```

Comment:
maybe add a test to enable/disable maintenance for the primary

Reply:
Done now.